From 46fb7bd6cf1fcb722de1b9454fe950d2d80c5656 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Tue, 09 Jul 2024 19:08:11 +0300
Subject: [PATCH] [FSW-7134] boot_desc: use dev spec descriptor parsing

For FS1 the descriptor structure will change significantly, so the
common parsing code of the descriptor used for G2/G3 can't be used
for FS1. As a preparation for the new descriptor structure, the
code parsing the descriptor is moved to device specific code, and in the
following patch the implementation for FS1 will be added.

Change-Id: Ib175cb64428e579c8c05ea9654e351f4282f34a5
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 25b675c..b1b7807 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -19,6 +19,8 @@
 #include <fs1_security.h>
 #include <status.h>
 #include <efuse.h>
+#include <hl_fw_upd_if.h>
+#include <boot_desc.h>
 #include "dev_uart.h"
 #include "progress_bits.h"
 
@@ -325,3 +327,19 @@
 
 	return freq_config;
 }
+
+bool dev_spec_img_desc_is_valid(__unused struct hl_fw_img_desc *img_desc)
+{
+	/* TODO: implement for FS1 if required */
+
+	return true;
+}
+
+uint32_t boot_desc_copy_desc(__unused struct boot_desc *desc,
+			     __unused struct hl_fw_img_desc *local_img_desc,
+			     __unused enum fw_post_reset_cmd upd_status)
+{
+	/* TODO: implement for FS1 */
+
+	return 0;
+}
diff --git a/common/boot_desc.c b/common/boot_desc.c
index 3fce0b1..3b0cd3a 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -29,10 +29,7 @@
 	    (img_desc->desc_ver == INV_VER2))
 		return false;
 
-	if (img_desc->idx >= INVALID_DESC_ID)
-		return false;
-
-	return true;
+	return dev_spec_img_desc_is_valid(img_desc);
 }
 
 static void boot_desc_update_ptrs(enum hl_fw_boot_ptr_imgs img)
@@ -160,29 +157,6 @@
 		return 1;
 
 	return 0;
-}
-
-static uint32_t boot_desc_copy_desc(struct boot_desc *desc,
-				    struct hl_fw_img_desc *local_img_desc,
-				    enum fw_post_reset_cmd upd_status)
-{
-	uint32_t idx;
-
-	idx = local_img_desc->idx;
-	if (upd_status == FLOPS_UPD_PENDING)
-		idx = (idx + 1) % 2;
-
-	desc->spi_offset = local_img_desc->offset[idx];
-	desc->img_size = local_img_desc->size[idx];
-	desc->loadaddr = ((uint64_t) local_img_desc->load_high[idx] << 32) |
-		local_img_desc->load_low[idx];
-	desc->desc_ver = local_img_desc->desc_ver;
-	desc->start_addr = ((uint64_t) local_img_desc->start_high[idx] << 32) |
-		local_img_desc->start_low[idx];
-
-	memcpy(desc->chksum, local_img_desc->chksum2[idx], ARRAY_SIZE(desc->chksum));
-
-	return idx;
 }
 
 static struct hl_fw_img_desc * boot_desc_img_desc_addr_calc(enum hl_fw_boot_ptr_imgs img,
diff --git a/common/boot_desc.h b/common/boot_desc.h
index 1028a05..87fa831 100644
--- a/common/boot_desc.h
+++ b/common/boot_desc.h
@@ -29,6 +29,9 @@
 uint32_t boot_desc_get_img_desc(enum hl_fw_boot_ptr_imgs img, int type,
 				struct boot_desc *desc, enum fw_post_reset_cmd upd_status);
 uint32_t boot_desc_param_get(enum hl_fw_boot_ptr_imgs img, struct hl_fw_boot_ptrs *boot_ptr);
+bool dev_spec_img_desc_is_valid(struct hl_fw_img_desc *img_desc);
+uint32_t boot_desc_copy_desc(struct boot_desc *desc, struct hl_fw_img_desc *local_img_desc,
+			     enum fw_post_reset_cmd upd_status);
 
 #endif /* _BOOT_DESC_H_ */
 
